### PR TITLE
FSharpWorkspace prototype

### PR DESCRIPTION
**Pull request by 0101**
## Summary

An F# workspace based on a dependency graph of files, projects, references. 

It's kind of an early prototype but whoever is interested can already have a look and maybe give some useful feedback. 

It can right now in VS-LSP deal with opening/closing/editing files, adding and removing files in projects, in-memory F# project references. 

## Public API so far

The user will create an instance of `FSharpWorkspace` and hold on to it. It's mutable and will be keeping the state of the workspace. 

They can add F# projects to it with:

- `AddCommandLineArgs(projectPath, outputPath, compilerArgs)`  
- That's it for now but we can potentially add other ways if people need it

They can update the projects if anything changes the same way. Just call `AddCommandLineArgs` with the new arguments.

 - Dependencies between projects are automatically discovered from the compiler args (`-r:` matching `outputPath`). So `outputPath` has to be provided correctly for in-memory references to work.  

They notify about files opening/closing/being edited by  `OpenFile`, `CloseFile`, `ChangeFile`. 

They can request `FSharpProjectSnapshot` for a given file and it will be lazily computed based on current state. The snapshot is still immutable and contains everything needed by `FSharpChecker` to work with the file. Parts of it may be lazily computed (e.g. if we're only interested in parsing it will not compute hash of references).

## Some technical details

To save some hash re-computation I decide to further (internally) split `ProjectSnapshot` into 
- `ProjectCore` = everything except files and (in-memory) references
- `ProjectBase` = `ProjectCore` + (in-memory) references
- `ProjectSnapshot` = `ProjectBase` + source files

I need help with names! 😂

I've been trying to come up with the simplest model for lazy "reactive" computation for the workspace. Right now it's the `DependencyGraph`. I don't love how it deals with different types of dependencies (DUs and "unpacking" them) but not sure how to solve it better without some sort of type-level programming 😅

### CODE REVIEW

File: src/Compiler/Service/FSharpProjectSnapshot.fs [hunk 1]
```diff
@@ -72,6 +72,13 @@ type FSharpFileSnapshot(FileName: string, Version: string, GetSource: unit -> Ta
     static member Create(fileName: string, version: string, getSource: unit -> Task<ISourceTextNew>) =
         FSharpFileSnapshot(fileName, version, getSource)
 
+    static member CreateFromString(filename: string, content: string) =
+        FSharpFileSnapshot(
+            filename,
+            Md5Hasher.hashString content |> Md5Hasher.toString,
```
Comment (vzarytovskii) [inline src/Compiler/Service/FSharpProjectSnapshot.fs:78]: Should probably eventually change to something more proper, can't ship with md5 hashing

**Review by vzarytovskii — APPROVED** [review]
Personally, I think we should just merge it and iterate further.

File: src/Compiler/Service/FSharpProjectSnapshot.fs (hunk 1 above)
Comment (0101) [reply src/Compiler/Service/FSharpProjectSnapshot.fs:78]: Most definitely. We should probably bring in some version of xxHash.

**Comment by 0101** [comment]
> Personally, I think we should just merge it and iterate further.

Yeah I just thought this might be a good point for people to start looking at it and poke holes or give advice on how to do it better.

File: src/Compiler/Service/FSharpProjectSnapshot.fs (hunk 1 above)
Comment (vzarytovskii) [reply src/Compiler/Service/FSharpProjectSnapshot.fs:78]: Yeah. That was my thinking.

**Comment by nojaf** [comment]
It would be nice if some NuGet packages could get attached to this PR (Upload artifact stuff in GitHub Actions). This would make it easier to play with in fsautocomplete for example.

**Comment by vzarytovskii** [comment]
> It would be nice if some NuGet packages could get attached to this PR (Upload artifact stuff in GitHub Actions). This would make it easier to play with in fsautocomplete for example.
> 
> 

Yeah we can do that, but probably won't be final, since types/packages/namespaces/etc can change.

**Comment by psfinaki** [comment]
I am still grasping the topic but...

> To save some hash re-computation I decide to further (internally) split ProjectSnapshot into
> - ProjectCore = everything except files and (in-memory) references
> - ProjectBase = ProjectCore + (in-memory) references
> - ProjectSnapshot = ProjectBase + source files

> I need help with names! 😂

How about ProjectProperties, ProjectPropertiesWithReferences and Project, respectively?

Base and Core and honestly very obscure suffixes. Better go long! :D

**Review by T-Gro — APPROVED** [review]

**Comment by 0101** [comment]
> 
> How about ProjectProperties, ProjectPropertiesWithReferences and Project, respectively?
> 
> Base and Core and honestly very obscure suffixes. Better go long! :D

`ProjectProperties` could work. Although it's quite similar to `ProjectOptions` which we already have (and which I'd like to deprecate at some point).

**Pull request merged**